compute_ctl: schema API endpoints

prepor opened the pull request:
## Problem

## Summary of changes

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

prepor commented:
I forgot about OpenAPI, working on it

tristan957 commented on compute_tools/src/schema.rs:
```diff
@@ -0,0 +1,116 @@
+        _ => {
+            let mut lines = stderr_reader.lines();
+            if let Some(line) = lines.next_line().await? {
+                if line.contains(&format!("FATAL:  database \"{}\" does not exist", dbname)) {
+                    return Err(SchemaDumpError::DatabaseDoesNotExist);
+                }
```
Do you know if this message is generated by the postgres server or the pg_dump client?

prepor replied:
I think it's generated by server and pg_dump just redirects it to stdout, but didn't find the exact codepath quickly.

tristan957 replied:
I was asking in case the server sends back a translated error message in the future. But we can cross the bridge later.

tristan957 commented on compute_tools/src/http/api.rs:
```diff
@@ -133,6 +137,38 @@ async fn routes(req: Request<Body>, compute: &Arc<ComputeNode>) -> Response<Body
+            };
+            info!("serving /schema/dump GET request with db_name: {db_name}",);
+            match schema_dump(compute, &db_name).await {
+                Ok(res) => Response::builder()
+                    .header(CONTENT_TYPE, "text/plain")
+                    .body(Body::wrap_stream(res))
+                    .unwrap(),
```
Maybe it would be good to add a `render_plain` function similar to the `render_json` function you added?

prepor replied:
Ok. Btw, we do not return content-type in other API responses, probably we need to fix it, but it's out of the scope of this PR, so I decided not to introduce these changes.

tristan957 replied:
Yep, that is definitely something that should be fixed. Can you create a ticket and assign it to me?

prepor replied:
done https://github.com/neondatabase/neon/issues/7629

tristan957 commented on compute_tools/src/http/openapi_spec.yaml:
```diff
@@ -68,6 +68,44 @@ paths:
+            text/plain:
+              schema:
+                type: string
+                description: Schema dump in SQL format.
```
add a newline after this.

tristan957 commented on compute_tools/src/http/openapi_spec.yaml:
```diff
@@ -68,6 +68,44 @@ paths:
+              schema:
+                $ref: "#/components/schemas/SchemaObjects"
+
+  /schema/dump:
```
I think `/schema/sql` or `/schema/ddl`  might be a better endpoint name.

prepor replied:
done

tristan957 commented on compute_tools/src/http/openapi_spec.yaml:
```diff
@@ -68,6 +68,45 @@ paths:
+        - Info
+      summary: Get schema dump
+      parameters:
+        - name: db_name
```
Would it be more consistent to just use `database` or do what Postgres does and use `dbname`?

prepor replied:
done, db_name -> database

tristan957 commented on compute_tools/src/schema.rs:
```diff
@@ -0,0 +1,116 @@
+    pg_helpers::{get_existing_dbs, get_existing_roles},
+};
+
+pub async fn get_schema_objects(compute: &Arc<ComputeNode>) -> anyhow::Result<SchemaObjects> {
```
It would be good to name the functions similarly

prepor replied:
done, schema_dump -> get_schema_ddl

tristan957 reviewed (commented):
Last couple of comments.

prepor commented on compute_tools/src/http/openapi_spec.yaml:
```diff
@@ -68,6 +68,45 @@ paths:
+    get:
+      tags:
+        - Info
+      summary: Get databases and roles in the schema.
```
In our case we — no, we always need all of this. I think retrieving this information is cheap enough to not complicate API.

ololobus commented on compute_tools/src/http/openapi_spec.yaml:
```diff
@@ -229,6 +268,67 @@ components:
+      properties:
+        name:
+          type: string
+        owner:
```
Let's also specify that we return these attributes
```
    #[serde(skip_deserializing, default)]
    pub restrict_conn: bool,
    #[serde(skip_deserializing, default)]
    pub invalid: bool,
```
with database (I think we should)

prepor replied:
done

ololobus commented on compute_tools/src/http/openapi_spec.yaml:
```diff
@@ -68,6 +68,45 @@ paths:
               schema:
                 $ref: "#/components/schemas/Info"
 
+  /schema/objects:
```
NIT: I don't think that roles and databases can be considered as schema, they are still records within some schema, so I'd actually separate these two routes into

```
/dbs_and_roles

DatabasesAndRoles: ...
Database: ...
Role: ...
GenericOption: ...
```

and

```
/schema
```

prepor replied:
> they are still records within some schema

so, they are part of the schema, aren't they? :)

ololobus replied:
Well, to me -- not, schema is schema and roles / dbs are records, effectively data, just in the shared catalog, but maybe I overthink it :)

IIRC, `pg_dump --schema-only` won't dump roles and dbs, because they are not considered as a part of schema. Schema is kinda db-level thing, while roles and dbs are above

I don't know, I was just confused seeing dbs / roles nearby schema

prepor replied:
@tristan957 WDYT?

tristan957 replied:
I think if Alexey's comment is right:

> IIRC, pg_dump --schema-only won't dump roles and dbs, because they are not considered as a part of schema. Schema is kinda db-level thing, while roles and dbs are above

Then we should move the `/objects` endpoint out of the /schema` endpoint.

prepor replied:
Ok, how about this:

I renamed `/schema/objects` to `catalog/objects` and `/schema/ddl` to `/catalog/database_ddl`. Does it look like more correct naming? WDYT?

ololobus replied:
Sorry, don't really want to do bikeshedding, but I'd
- replace `/catalog/database_ddl` with just `/schema` or `/database_schema`. It's more straightforward and clean. I think in Postgres catalog or shared catalog or pg_catalog is usually used to refer shared tables https://www.postgresql.org/docs/current/catalogs.html and it's not what we get here
- `catalog/objects` is probably OK, but sounds a bit too broad, so I really don't see any cons in just naming it more specific like `/dbs_and_roles`, do you?

prepor replied:
> I think in Postgres catalog or shared catalog or pg_catalog is usually used to refer shared tables https://www.postgresql.org/docs/current/catalogs.html and it's not what we get here

yes! And pg_dump --schema-only is basically a utility that transforms pg_catalog entities into DDL statements, isn't it? 

@tristan957 please, make your choice :)

tristan957 replied:
I'll defer to Alexey :). Thanks for your work on this thus far.

ololobus reviewed (commented):
LGTM, but left some comments

As this PR goes first and we don't have any e2e tests merged yet, let's create a simple python test in the `neon` repo? E.g. start endpoint, create new role and db; call both APIs and do basic sanity checks that result makes sense

prepor commented on compute_tools/src/http/openapi_spec.yaml:
```diff
@@ -68,6 +68,45 @@ paths:
+      description: Get schema dump in SQL format.
+      operationId: getSchemaDump
+      responses:
+        200:
```
oops, thank you, I forgot about it, fixed

prepor commented:
> As this PR goes first and we don't have any e2e tests merged yet, let's create a simple python test in the neon repo?

It doesn't feel right for me to introduce e2e tests into the project as part of this PR. Not sure that everyone will be happy with python as testing tool, for example. But I would be happy to add e2e tests for these cases after e2e toolbox is introduced for the project.

BTW, we do have a simple e2e test for schema dump as part of our cloud e2e tests suite.

ololobus commented:
> It doesn't feel right for me to introduce e2e tests into the project as part of this PR. Not sure that everyone will be happy with python as testing tool, for example. But I would be happy to add e2e tests for these cases after e2e toolbox is introduced for the project.

But there is python test tooling in this repo. For example, test case for related area https://github.com/neondatabase/neon/blob/55ba885f6be431b2b694bca4d95d7a40f4bda98a/test_runner/regress/test_ddl_forwarding.py#L291

I don't quite get, why we cannot write a basic test for this new code using existing tooling in the repo

prepor commented:
@ololobus I'm sorry, I misunderstood your previous comment. I added tests for these new API endpoints. I haven't found any existing Compute API tests, so I also added a simple wrapper for Compute's HTTP API

tristan957 commented on test_runner/fixtures/endpoint/http.py:
```diff
@@ -0,0 +1,23 @@
+        res.raise_for_status()
+        return res.json()
+
+    def catalog_schema_ddl(self, database: str):
```
rename the function to catalog_datadbase_ddl to match the endpoint.
